src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CreateNewDataFilesCommand.java

```diff
@@ -323,156 +316,104 @@ public CreateDataFileResult execute(CommandContext ctxt) throws CommandException
+                                if (combinedUnzippedFileSize > storageQuotaLimit) {
+                                    //throw new FileExceedsStorageQuotaException(MessageFormat.format(BundleUtil.getStringFromBundle("file.addreplace.error.quota_exceeded"), bytesToHumanReadable(combinedUnzippedFileSize), bytesToHumanReadable(storageQuotaLimit)));
+                                    // change of plans: if the unzipped content inside exceeds the remaining quota,
+                                    // we reject the upload outright, rather than accepting the zip
+                                    // file as is.
+                                    throw new CommandExecutionException(MessageFormat.format(BundleUtil.getStringFromBundle("file.addreplace.error.unzipped.quota_exceeded"), bytesToHumanReadable(storageQuotaLimit)), this);
                                 }
                             }
                         }
                     }
-                    
+
                     // OK we're still here - that means we can proceed unzipping. 
                     
-                    // Close the ZipFile, re-open as ZipInputStream: 
-                    zipFile.close(); 
                     // reset:
                     combinedUnzippedFileSize = 0L;
 
-                    if (charset != null) {
-                        unZippedIn = new ZipInputStream(new FileInputStream(tempFile.toFile()), charset);
-                    } else {
-                        unZippedIn = new ZipInputStream(new FileInputStream(tempFile.toFile()));
-                    }
-
-                    while (true) {
-                        try {
-                            zipEntry = unZippedIn.getNextEntry();
-                        } catch (IllegalArgumentException iaex) {
-                            // Note:
-                            // ZipInputStream documentation doesn't even mention that
-                            // getNextEntry() throws an IllegalArgumentException!
-                            // but that's what happens if the file name of the next
-                            // entry is not valid in the current CharSet.
-                            // -- L.A.
-                            warningMessage = "Failed to unpack Zip file. (Unknown Character Set used in a file name?) Saving the file as is.";
-                            logger.warning(warningMessage);
-                            throw new IOException();
-                        }
-
-                        if (zipEntry == null) {
-                            break;
-                        }
-                        // Note that some zip entries may be directories - we
-                        // simply skip them:
-
-                        if (!zipEntry.isDirectory()) {
+                    try (var zipFile = openZipFile(tempFile, charset)) {
+                        for (var entry : filteredZipEntries(zipFile)) {
```

Comment:
Can you avoid re opening the zip and getting entries since you already have them?

Reply:
Thought I tried and failed. Thanks for making me retry.